[TF:MLIR] Canonicalize ShapeNOp with partial static input shape

WindQAQ opened the pull request:
Canonicalize ShapeNOp by replacing corresponding output with static input shape.

WindQAQ committed:
Canonicalize ShapeNOp with partial static input shape

Simplify variable

WindQAQ commented:
/cc @smit-hinsu for visibility.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,42 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+    SmallVector<Value, 4> results;
+    for (Value input : op.getOperands()) {
+      Value shape;
+      if (OpFoldResult result = ConvertShapeToAttr(input.getType(), width)) {
```
ConvertShapeToAttr return type is Attribute.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,42 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+        shape =
+            rewriter.create<TF::ConstOp>(op.getLoc(), result.get<Attribute>());
+      } else {
+        shape = rewriter.create<TF::ShapeOp>(op.getLoc(), input, use32Bit);
```
Instead of creating individual Shape ops, we should create a single ShapeN op with all the dynamic inputs.

WindQAQ replied:
Updated! A quick followup here. Should we canonicalize `ShapeNOp` to `ShapeOp` when n = 1?

smit-hinsu replied:
Yes, that's a good idea.

WindQAQ replied:
> Updated! A quick followup here. Should we canonicalize `ShapeNOp` to `ShapeOp` when n = 1?

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,42 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+  using OpRewritePattern<ShapeNOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(ShapeNOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.getNumOperands() == 0) return success();
```
Add a comment that folder handles this case.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/tests/constant-fold.mlir:
```diff
@@ -89,16 +89,18 @@ func @testEmptybf16() -> (tensor<5xbf16>) {
 
-  // CHECK: tf.ShapeN
-  %1:2 = "tf.ShapeN"(%arg1, %arg2) : (tensor<1x32x32x16xf32>, tensor<*xf32>) -> (tensor<4xi64>, tensor<?xi64>)
+  // CHECK: %[[SHAPE3:.*]] = "tf.Const"() {value = dense<[1, 32, 32]> : tensor<3xi64>}
```
nit: this is a bit convoluted now. SHAPE3 comes before SHAPE2 and SHAPE1 ops are de-duped which is not related to our canonicalizer. Let's split out this op in a separate function.

smit-hinsu reviewed (changes requested):
Thanks for the contributions!

WindQAQ committed:
Create a ShapeNOp instead of multiple ShapeOps

Use getType directly

Create ShapeOp instead of ShapeNOp when there is only one dynamic input

Run clang-format

Use plural variable name

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,60 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+                                PatternRewriter &rewriter) const override {
+    // ShapeNOp::fold handles this case.
+    if (op.getNumOperands() == 0) return success();
+    int width = op.getType(0)
```
getelementtypeorself helper function will help shorten this.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,60 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+                    .getIntOrFloatBitWidth();
+
+    SmallVector<Value, 4> results(op.getNumOperands());
+    SmallVector<int64_t, 4> dynamic_index;
```
nit: rename to dynamic_indices and dynamic_inputs

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,60 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+      return failure();
+    }
+
+    if (!dynamic_input.empty()) {
```
Now that we handle all static shaped operands case here, we can drop the folder.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc:
```diff
@@ -947,9 +947,65 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
+
+    // Create a ShapeOp when there is only one dynamic input.
+    // Or create a ShapeNOp when there are two or more dynamic inputs.
+    if (dynamic_input.size() == 1) {
```
Having a separate pattern for ShapeN to Shape conversion to prefer Shape op is useful but not necessary. I don't see big advantage to condition it here.

WindQAQ committed:
Separate ShapeNToShape canonicalization pattern

Remove folder

Add tests

Use getElementTypeOrSelf utility

smit-hinsu reviewed (approved):
Good tests!
